src/Avalonia.Base/Media/GeometryBuilder.cs

```diff
@@ -0,0 +1,610 @@
+                double v = (fLeftTop) / (fLeftTop + fRightTop) * boundRect.Width;
+                topLeftX = v;
+                topRightX = v;
+            }
+            // right edge
+            if (rightTopY > rightBottomY)
+            {
+                double v = (fTopRight) / (fTopRight + fBottomRight) * boundRect.Height;
+                rightTopY = v;
+                rightBottomY = v;
+            }
+            // bottom edge
+            if (bottomRightX < bottomLeftX)
+            {
+                double v = (fLeftBottom) / (fLeftBottom + fRightBottom) * boundRect.Width;
+                bottomRightX = v;
+                bottomLeftX = v;
+            }
+            // left edge
+            if (leftBottomY < leftTopY)
+            {
+                double v = (fTopLeft) / (fTopLeft + fBottomLeft) * boundRect.Height;
+                leftBottomY = v;
+                leftTopY = v;
+            }
+
+            // The above code does all calculations without taking into consideration X/Y absolute position.
+            // In WinUI, this is compensated for in DrawRoundedCornersRectangle(); however, we do this here directly
+            // when the final keypoints are being created.
+            var keypoints = new RoundedRectKeypoints();
+            keypoints.TopLeft = new Point(
+                boundRect.X + topLeftX,
+                boundRect.Y + topLeftY);
+            keypoints.TopRight = new Point(
+                boundRect.X + topRightX,
+                boundRect.Y + topRightY);
+
+            keypoints.RightTop = new Point(
+                boundRect.X + rightTopX,
+                boundRect.Y + rightTopY);
+            keypoints.RightBottom = new Point(
+                boundRect.X + rightBottomX,
+                boundRect.Y + rightBottomY);
+
+            keypoints.BottomRight = new Point(
+                boundRect.X + bottomRightX,
+                boundRect.Y + bottomRightY);
+            keypoints.BottomLeft = new Point(
+                boundRect.X + bottomLeftX,
+                boundRect.Y + bottomLeftY);
+
+            keypoints.LeftBottom = new Point(
+                boundRect.X + leftBottomX,
+                boundRect.Y + leftBottomY);
+            keypoints.LeftTop = new Point(
+                boundRect.X + leftTopX,
+                boundRect.Y + leftTopY);
+
+            return keypoints;
+        }
+
+        /// <summary>
+        /// Represents the keypoints of a rounded rectangle.
+        /// These keypoints can be shared between methods and turned into geometry.
+        /// </summary>
+        /// <remarks>
+        /// A rounded rectangle is the base geometric shape used when drawing borders.
+        /// It is a superset of a simple rectangle (which has corner radii set to zero).
+        /// These keypoints can be combined together to produce geometries for both background
+        /// and border elements.
+        /// </remarks>
+        internal struct RoundedRectKeypoints
```

Comment:
Note that this was switched from a class to a struct for theoretical performance reasons. However, it is now mutable.

The original class was in BorderRenderHelper called `BorderGeometryKeypoints`.

Reply:
Hm, it's an interesting one. Changing it to struct would play well, if this struct was only used in a scope of the single method. But since it's a return type of the method, and also an argument of another, it might get copied way too much.

Reply:
Yea, point taken. There are a lot of members in `RoundedRectKeypoints`. I was trying to avoid the overhead of a class here though.

Since this is an internal API I see no problem in passing by `ref` to the related methods to avoid the copy. Then we get both optimizations. Seems appropriate considering this is fairly low level and graphics-related.